purchase_sale_inter_company/models/purchase_order.py

```diff
@@ -175,14 +194,23 @@ def _prepare_sale_order_line_data(self, purchase_line, dest_company, sale_order)
         return new_line._convert_to_write(new_line._cache)
 
     def button_cancel(self):
-        sale_orders = (
-            self.env["sale.order"]
-            .sudo()
-            .search([("auto_purchase_order_id", "in", self.ids)])
-        )
-        for so in sale_orders:
-            if so.state not in ["draft", "sent", "cancel"]:
-                raise UserError(_("You can't cancel an order that is %s") % so.state)
-        sale_orders.action_cancel()
-        self.write({"partner_ref": False})
+        if self.intercompany_sale_id:
+            if self.intercompany_sale_id.state not in ["draft", "sent", "cancel"]:
+                raise UserError(
+                    _(
+                        "You can not cancel your purchase order %s.\n"
+                        "The sale order %s at your supplier %s that is linked "
+                        "to your purchase order, is in the state %s.\n"
+                        "In order to cancel your purchase order, you must first "
+                        "ask your supplier to cancel his sale order."
+                    )
+                    % (
+                        self.name,
+                        self.intercompany_sale_id.name,
+                        self.partner_id.name,
+                        self.intercompany_sale_id.state,
+                    )
+                )
+            self.intercompany_sale_id.sudo().action_cancel()
+            self.write({"partner_ref": False})
```

Comment:
**question**: why delete the `partner_ref`? The PO is still the same even if it was cancelled